src/ucp/core/ucp_request.c

```diff
@@ -461,51 +400,73 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_request_memory_reg,
          * Forget about what user provided and register what we need.
          */
         ucp_trace_req(req, "mds 0x%" PRIx64 " not registered - drop user memh",
-                      md_map & ~state->dt.contig.md_map);
+                      md_map & ~state->dt.contig.memh->md_map);
         req->flags             &= ~UCP_REQUEST_FLAG_USER_MEMH;
-        state->dt.contig.md_map = 0;
+        state->dt.contig.memh   = NULL;
```

Comment:
seems we can remove lines 388-406 now. we **can** mix user-provided memh with internal registration - by calling ucp_memh_update